Updated Server Logger and OnSession functions

brnsampson opened the pull request:
Logger is a function that should return a per-request *slog.Logger

OnSession changed to remove Session and Subscription references
- Accepts (w http.ResponseWriter, r *http.Request)
- Returns (topics []string, allowed bool)

I updated the complex example and server_test.go test to account for the changes as well.

I tried to avoid altering any behavior, with the exception that if a session was accepted by the user but an empty or nil topic slice is returned we just use the default slice instead of raising a panic.

brnsampson commented:
I just realized that I didn't address the user returning (nil, true) from OnSession. One sec.

tmaxmax commented on cmd/complex/main.go:
```diff
@@ -25,28 +26,41 @@ func newSSE() *sse.Server {
+						topicMetrics,
+					)
+
+					// NOTE: if you are returning false to reject the subscription you must also return
+					// your own error code!
+					w.WriteHeader(http.StatusBadRequest)
```
Technically this is not true – if a `ServeHTTP` ends without a response code, a `200 OK` will be implicitly written.

Let's maybe change this to something like:
```suggestion
					// NOTE: if you are returning false to reject the subscription we recommend
					// writing your own response code. Clients will receive a confusing 200 OK otherwise.
					w.WriteHeader(http.StatusBadRequest)
```

brnsampson replied:
✅

tmaxmax commented on cmd/complex/logger.go:
```diff
@@ -2,77 +2,46 @@ package main
+
+// getLogger retrieves the request-specific logger from a request's context. This is
+// similar to how existing packages like https://github.com/go-chi/httplog work.
+func getLogger(r *http.Request) (*slog.Logger, error) {
+	logger, ok := r.Context().Value(loggerCtxKey{}).(*slog.Logger)
+	if !ok {
+		return nil, errors.New("logger not available on request")
+	}
+
+	return logger, nil
+}
+
+// setupLogger just initializes the global logger in whatever way you might prefer. We are
+// keeping it simple for this example.
+func setupLogger() {
+	h := slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelDebug})
+	logger := slog.New(h)
+	slog.SetDefault(logger)
+}
```
Given that `getLogger` is used in just one place let's remove boilerplate and have this function fit for `sse.Server` directly.

Also, do we need a global logger? This code doesn't call the global logger at all. What if we just initialize that logger in `main` and pass it where necessary? We could also have `withLogger` take in the base logger and not rely on `slog.Default()`. No globals should make code easier to follow.

```suggestion
func getLogger(r *http.Request) *slog.Logger {
	return r.Context().Value(loggerCtxKey{}).(*slog.Logger)
}
```

brnsampson replied:
In the withLogger function a new logger handle is created from slog.Default(), which returns a pointer to the global logger. Looking at it now, I was planning on adding a log message to the OnSession function which would also use the default logger, since that seems like a likely thing for someone to do in practice, but left it out because it seemed like an unneccesary addition.

I could put all this stuff just in main if you prefer? I just split it up this way to keep the most 1-to-1 translation of what was already there.

brnsampson replied:
Oh, I also did add a small thing to demonstrate how to get the http.Server error logs at the same time. That also utilizes the default logger to create a *log.Logger that http.Server.ErrorLog can use which ties the whole logging story together with a single handler.

brnsampson replied:
If it's okay to change the structure of the example more significantly, I think what I would tend to do normally would be to make getLogger a closure like `getLogger(*slog.Logger) func(r *http.Request) *slog.Logger`. I would then explicitly pass a *slog.Logger to the newSSE function and create a logger object in main() to pass around.

I don't personally like globals, so that would be my preference anyways.

tmaxmax replied:
But if `getLogger` returns a closure over the given logger, does it still retrieve a logger from the request context? If it does, then passing a logger through the closure doesn't make sense, as it will not be used.

The `http.Server.ErrorLog` thing is nice, didn't know about it!

brnsampson replied:
oh sorry, I meant withLogger could be a closure. That way main() could instantiate a new logger and pass it to withLogger to create a `func(h http.Handler) http.Handler` middleware that creates the per-request loggers from there instead of the global logger.

tmaxmax replied:
Yes, that's also how I thought of it, let's do it this way.

brnsampson replied:
ok, take a look now

tmaxmax commented on server.go:
```diff
@@ -176,26 +153,28 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			l.Warn("sse: invalid subscription")
 		}
+
+		// NOTE: the user is expected to return an appropriate error code in OnSession.
```
Can probably remove this
```suggestion
```

brnsampson replied:
ah, that was a reminder to myself until I finished that section. I just forgot to delete it

brnsampson replied:
✅

tmaxmax commented on server.go:
```diff
@@ -125,17 +96,20 @@ type Server struct {
+	// a new subscription for the connection and events will be sent
+	// to this client, otherwise the request will be ended.
+	//
+	// Note that you must return any error codes to the client using
+	// the http.ResponseWriter yourself based on the reason for rejection.
```
We should change this comment to be like the one from the example.

tmaxmax commented on server.go:
```diff
@@ -229,20 +208,17 @@ func (s *Server) init() {
+	sub := Subscription{Client: sess, LastEventID: sess.LastEventID, Topics: defaultTopicSlice}
 	if s.OnSession != nil {
-		sub, ok := s.OnSession(sess)
-		if ok && len(sub.Topics) == 0 {
-			panic("go-sse: session handlers should return a Subscription to at least 1 topic")
+		topics, ok := s.OnSession(sess.Res, sess.Req)
+		if ok && len(topics) > 0 {
+			sub.Topics = topics
 		}
 
 		return sub, ok
 	}
```
looks nice

tmaxmax commented on server_test.go:
```diff
@@ -155,13 +134,18 @@ func TestServer_ServeHTTP(t *testing.T) {
+	tests.Equal(
+		t,
+		sb.String(),
+		"level=INFO msg=\"sse: starting new session\"\nlevel=INFO msg=\"sse: subscribing session\" topics=[] lastEventID=5\nlevel=INFO msg=\"sse: session ended\"\n",
```
Are the log messages compatible with the `slog` output format?

Maybe we should make the `mockHandler` smarter: it could collect all `slog.Record` objects in a slice and then we could compare the records accumulated by the `mockHandler` with an expected list.

This way we don't depend on how `slog` formats the log output.

brnsampson replied:
yeah, I had to add that hack to remove the timestamp from the output, but after that it looked identical and the tests passed 🤷

tmaxmax commented on server_test.go:
```diff
@@ -125,7 +100,12 @@ func TestServer_ShutdownPublish(t *testing.T) {
 	_ = s.Publish(&sse.Message{})
 	tests.Expect(t, p.Published, "Publish wasn't called")
-	tests.DeepEqual(t, []any{*p.Pub, p.PubTopics}, []any{sse.Message{}, []string{sse.DefaultTopic}}, "incorrect message")
+	tests.DeepEqual(
+		t,
+		[]any{*p.Pub, p.PubTopics},
+		[]any{sse.Message{}, []string{sse.DefaultTopic}},
+		"incorrect message",
+	)
```
Can we remove these formatting changes? Most of them are not related to the change set in this PR anyway.

Stylistically I also prefer not breaking lines in most cases, even if they are very long.

brnsampson replied:
This was the golines LSP doing it's thing again. I'll see if there is some kind of way to temporarily disable it.

brnsampson replied:
I think I got it. Let me know if there are other formatting problems that it caused

tmaxmax reviewed (commented):
We should also modify the CHANGELOG accordingly.

tmaxmax commented on cmd/complex/logger.go:
```diff
@@ -2,77 +2,46 @@ package main
-
 type loggerCtxKey struct{}
 
+// withLogger is an http middleware that generates a logger with request-specific fields
+// added to it and attaches it to the request context for later retrieval with getLogger().
+// This is simmilar to how existing packages like https://github.com/go-chi/httplog works.
```
I'd remove links to external packages. They look like endorsements to me and I don't think that's appropriate. Comments should also use correct initialism for acronyms ("HTTP", not "http").
```suggestion
// withLogger is an HTTP middleware that creates a logger with request-specific fields
// added to it and attaches it to the request context for later retrieval with getLogger().
```

brnsampson replied:
This is one of those "specific beats general" cases in my mind, since it's the HTTP protocol but I'm referencing the more-or-less standard pattern for middleware for the stdlib `http` package.

I did remove the specific reference for a more general reference to 3rd party libraries though.

tmaxmax replied:
Oh, I see what you mean – maybe in that case we could write "net/http" or "stdlib-compatible" instead of just "http"? The latter is too easy to brush off as just the HTTP protocol, as I initially did.

Pertaining to the 3rd party references, I'd probably be a bit more specific on how they're similar and rephrase the comments to:
- for `withLogger`: "Third party logging packages may offer similar middlewares to add a logger to the request or maybe just a helper to add a logger to context; in the second case you can build your own middleware function around it, similar to this one."
- for `getLogger`: "Third party logging packages may offer similar helpers for retrieving a logger from a request or a context."

brnsampson replied:
✅

tmaxmax commented on cmd/complex/main.go:
```diff
@@ -25,28 +26,41 @@ func newSSE() *sse.Server {
-		Logger:   logger{log.New(os.Stderr, "", 0)},
-		OnSession: func(s *sse.Session) (sse.Subscription, bool) {
-			topics := s.Req.URL.Query()["topic"]
+		Logger:   logFunc,
```
Though having a comment hinting at external helper usage is helpful. I'd probably add that comment here, as this knowledge is most relevant when initializing the `sse.Server`.
```suggestion
		// If you're using a third-party logging library and have ensured that this library
		// adds the logger to the request context, use here its own helper function
		// to retrieve a logger from context.
		Logger:   getLogger,
```
I've also changed `logFunc` to `getLogger`, assuming the other proposed change to the signature of `getLogger`.

brnsampson replied:
I added the comment and just made the logFunc inline in the struct initialization. It seems totally readable that way and is closer to what was originally there.

tmaxmax commented on server.go:
```diff
@@ -125,17 +96,20 @@ type Server struct {
+	// a new subscription for the connection and events will be sent
+	// to this client, otherwise the request will be ended.
+	//
+	// Note that you must return any error codes to the client using
+	// the http.ResponseWriter yourself based on the reason for rejection.
```
Given that it's incorrect to specify that one _must_ do it, let's rephrase it as follows:
```suggestion
	// Note that OnSession can write the HTTP response code itself, if something other
	// than the implicit 200 OK is desired. This is especially helpful when refusing sessions –
	// if OnSession does not write a response code, clients will receive a confusing 200 OK.
```

brnsampson replied:
✅

tmaxmax commented:
Partially related to this, I've just realised something: could you also remove [this line of code](https://github.com/tmaxmax/go-sse/blob/3412f19165d0687aab48627c00ed5f773babdfb0/session.go#L69) and add the following line in the CHANGELOG under the "Fixed" section:
```
- `sse.Session` doesn't write the header explicitly anymore. This would cause a `http: superfluous response.WriteHeader call` warning being logged when `sse.Server.OnSession` writes a response code itself when accepting a session. The change was initially introduced to remove the warning for users of certain external libraries (see #41) but this is the issue of the external library, not of `go-sse`. If you encounter this warning when using an external library, write the response code yourself in the HTTP handler before subscribing the `sse.Session`, as described in the linked discussion.
```

brnsampson commented:
Sure, I'm assuming this should be under 0.11.0 in the changelog?

tmaxmax commented:
Let's write "Unreleased" for now as a placeholder for the version. I write the version only when effectively releasing it.

tmaxmax commented on CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+- Server.Logger is now of type `func(r *http.Request) *slog.Logger` instead of `sse.Logger`
+- Server.OnSession signature changed from `func(s *sse.Session) (Subscription, bool)` to `func(w http.ResponseWriter, r
+*http.Request) (topics []string, accepted bool)`
+- complex example updated to demonstrate new behavior
```
That's technically not a change to the library itself so let's not cover it in the changelog.
```suggestion
```

brnsampson commented:
Okay, I think most of the comment changes have been addressed.

If you are fine with me changing the structure of the example a little more I would probably lean towards eliminating the default logger, inlining the getLogger and setupLogger functions to remove them, and just putting withLogger and loggerCtxKey into main since they are not very long.

I'll have to poke around for a minute to see how to disable some formatters in my editor to change those line breaks. Currently the formmaters just run on save so whenever I try to change it they get re-formatted when I close the file. I think it also made some very small formatting changes to CHANGELOG.md, but I think that may just have been whitespace related.

tmaxmax commented on CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
 
+## [Unreleased] - 2025-02-02
+
+Logging and Server request handling revamped to hopefully be more familiar interfaces for users.
```
```suggestion
The `sse.Server` logging and session handling were revamped to have more familiar, more flexible and less error prone interfaces for users.
```

tmaxmax commented on CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+
+### Removed
+
+- Logger and LogLevel enum have been removed.
```
Just some backticks:
```suggestion
- `Logger` and `LogLevel` enum have been removed. `Server.Logger` has transitioned to the standard `slog` library for better compatibility with the ecosystem
```

tmaxmax commented on CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+
+### Changed
+
+- Server.Logger is now of type `func(r *http.Request) *slog.Logger` instead of `sse.Logger`
```
Backticks and some more details:
```suggestion
- `Server.Logger` is now of type `func(r *http.Request) *slog.Logger` instead of `sse.Logger` – it is possible to customize the logger on a per-request basis, by for example retrieving it from the context. 
```

tmaxmax commented on CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+### Changed
+
+- Server.Logger is now of type `func(r *http.Request) *slog.Logger` instead of `sse.Logger`
+- Server.OnSession signature changed from `func(s *sse.Session) (Subscription, bool)` to `func(w http.ResponseWriter, r
+*http.Request) (topics []string, accepted bool)`
```
Backticks and more detail
```suggestion
- `Server.OnSession` signature changed from `func(s *Session) (Subscription, bool)` to `func(w http.ResponseWriter, r *http.Request) (topics []string, accepted bool)` – its initial role was to essentially just provide the topics, so the need to fiddle with `Session` and `Subscription` was redundant anyway
```

tmaxmax commented on CHANGELOG.md:
```diff
@@ -57,6 +76,7 @@ This version removes all external dependencies of `go-sse`. All our bugs are bel
 
 - `Server.Logger` is now of a new type: the `Logger` interface. The dependency on x/exp/slog is removed. This opens up the possibility to adapt any existing logger to be usable with `Server`.
 - The default backoff behavior has changed. The _previous_ defaults map to the new `Backoff` configuration as follows:
+
```
Can we remove these formatting changes or include them in a separate commit?

brnsampson replied:
✅ removed formatting changes

tmaxmax commented on server.go:
```diff
@@ -125,17 +96,22 @@ type Server struct {
+	// should be accepted or not. If it is true, the Provider will receive
+	// a new subscription for the connection and events will be sent
+	// to this client, otherwise the request will be ended.
+	//
+	//
```
Dumb nitpick but here we are
```suggestion
	//
```

tmaxmax reviewed (commented):
Sorry for the barrage of comments on the changelog – I haven't stated the expectations clearly. You should be able to just commit those suggestions from the GitHub UI and now waste any time with it.

tmaxmax commented on cmd/complex/main.go:
```diff
@@ -27,26 +28,39 @@ func newSSE() *sse.Server {
-					fmt.Fprintf(s.Res, "invalid topic %q; supported are %q, %q", topic, topicRandomNumbers, topicMetrics)
-					s.Res.WriteHeader(http.StatusBadRequest)
-					return sse.Subscription{}, false
+					fmt.Fprintf(
+						w,
+						"invalid topic %q; supported are %q, %q",
+						topic,
+						topicRandomNumbers,
+						topicMetrics,
+					)
```
I'd prefer this on a single line
```suggestion
					fmt.Fprintf(w, "invalid topic %q; supported are %q, %q", topic, topicRandomNumbers, topicMetrics)
```

tmaxmax commented on cmd/complex/main.go:
```diff
@@ -173,3 +193,35 @@ func generateRandomNumbers() *sse.Message {
+	}
+}
+
+// getLogger retrieves the request-specific logger from a request's context. This is
+// similar to how existing per-request http logging libraries work, just very simplified.
+func getLogger(r *http.Request) (*slog.Logger, error) {
```
I'd honestly remove the comment and take just a `context.Context`:
```suggestion
func getLogger(ctx context.Context) (*slog.Logger, error) {
```
This way we don't really have to explain anything – not only because it's self-explanatory that it takes a logger from a context, but also because this function is wrapped when initializing the server and there's already a comment there that about wrapping third party APIs.

Possibly also get rid of the error return, to keep things simple?

brnsampson replied:
✅

tmaxmax commented:
I've committed the suggested CHANGELOG adjustments and other nitpicks myself, to not waste time on those. Open comments left:
- [a discussion on the documentation of `withLogger`](https://github.com/tmaxmax/go-sse/pull/48#discussion_r1938762017)
- [some CHANGELOG formatting issues](https://github.com/tmaxmax/go-sse/pull/48#discussion_r1938746480)
- [discussion on `getLogger`](https://github.com/tmaxmax/go-sse/pull/48#discussion_r1938765552)

Apart from that everything's good to go.

tmaxmax commented on cmd/complex/main.go:
```diff
@@ -27,26 +28,33 @@ func newSSE() *sse.Server {
-			return sse.Subscription{
-				Client:      s,
-				LastEventID: s.LastEventID,
-				Topics:      append(topics, sse.DefaultTopic), // the shutdown message is sent on the default topic
```
Let's not delete this comment here.

brnsampson replied:
✅

brnsampson commented:
Apologies for the delay, we had some cat health drama that kept me busy. It was also far, far more involved to track down how to disable a builtin formatter for one file than I expected for some reason.

I think I got to all the comments, so let me know if there is anything else. I don't mind rebasing the whole branch into a single commit or if you want to squash and merge that's fine too.

tmaxmax commented:
Hi, everything looks great! Don't worry about any delay, code's not going anywhere. And, as always, family first 😌

Rebasing would be perfect! There are also some minor linter issues to fix. For one of them just add the value `hugeParam` in `.golangci.yml` in the `linter-settings.gocritic.disabled-checks` list.

brnsampson commented:
I haven't been able to get golangci-lint to work locally so I wasn't able to verify that it will run cleanly, but I added the exclusion and squashed everything down to 1 commit

tmaxmax commented on server_test.go:
```diff
@@ -210,10 +184,10 @@ func TestServer_OnSession(t *testing.T) {
-				http.Error(s.Res, "this is invalid", http.StatusBadRequest)
-				return sse.Subscription{}, false
+			Logger:   mockLogFunc(sb),
+			OnSession: func(w http.ResponseWriter, r *http.Request) ([]string, bool) {
```
```suggestion
			OnSession: func(w http.ResponseWriter, _ *http.Request) ([]string, bool) {
```

tmaxmax commented on server_test.go:
```diff
@@ -73,48 +74,22 @@ func newMockProvider(tb testing.TB, subErr error) *mockProvider {
-	}
-	if id, ok := data["lastEventID"]; ok {
-		output += fmt.Sprintf(" lastEventID=%s", quoteString(id.(sse.EventID).String()))
+func mockLogFunc(w io.Writer) func(r *http.Request) *slog.Logger {
```
```suggestion
func mockLogFunc(w io.Writer) func(*http.Request) *slog.Logger {
```

tmaxmax reviewed (commented):
these were the other 2 linter comments, i believe

tmaxmax commented on server_test.go:
```diff
@@ -73,48 +74,22 @@ func newMockProvider(tb testing.TB, subErr error) *mockProvider {
+func mockLogFunc(w io.Writer) func(*http.Request) *slog.Logger {
+	h := slog.NewTextHandler(w, nil)
+	mockH := mockHandler{h}
+	return func(r *http.Request) *slog.Logger {
```
```suggestion
	return func(*http.Request) *slog.Logger {
```

brnsampson commented:
Sorry I had to disappear again at the end, it has been a crazy week. Thanks for wrapping it up at the end!
